File: src/test/java/ru/jerael/booktracker/backend/web/controller/AuthControllerTest.java
package ru.jerael.booktracker.backend.web.controller;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest;
import org.springframework.context.annotation.Import;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.assertj.MockMvcTester;
import ru.jerael.booktracker.backend.domain.exception.factory.UserExceptionFactory;
import ru.jerael.booktracker.backend.domain.model.auth.ConfirmRegistration;
import ru.jerael.booktracker.backend.domain.model.auth.TokenPair;
import ru.jerael.booktracker.backend.domain.model.user.UserCreation;
import ru.jerael.booktracker.backend.domain.model.user.UserCreationResult;
import ru.jerael.booktracker.backend.domain.usecase.auth.ConfirmRegistrationUseCase;
import ru.jerael.booktracker.backend.domain.usecase.user.CreateUserUseCase;
import ru.jerael.booktracker.backend.web.config.WebProperties;
import ru.jerael.booktracker.backend.web.dto.auth.AuthResponse;
import ru.jerael.booktracker.backend.web.dto.auth.ConfirmRegistrationRequest;
import ru.jerael.booktracker.backend.web.dto.user.UserCreationRequest;
import ru.jerael.booktracker.backend.web.dto.user.UserCreationResponse;
import ru.jerael.booktracker.backend.web.mapper.AuthWebMapper;
import ru.jerael.booktracker.backend.web.mapper.UserWebMapper;
import tools.jackson.databind.ObjectMapper;
import java.time.Instant;
import java.util.UUID;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

@WebMvcTest(AuthController.class)
@Import({UserWebMapper.class, AuthWebMapper.class})
class AuthControllerTest {

    @Autowired
    private MockMvcTester mockMvcTester;

    @MockitoBean
    private WebProperties webProperties;

    @Autowired
    private ObjectMapper objectMapper;

    @MockitoBean
    private CreateUserUseCase createUserUseCase;

    @MockitoBean
    private ConfirmRegistrationUseCase confirmRegistrationUseCase;

    private final String email = "test@example.com";
    private final String password = "Password123!";
    private final UUID userId = UUID.fromString("ee39af7a-a073-4473-878a-1aae34e98bb7");
    private final Instant expiresAt = Instant.ofEpochMilli(1771249999347L);
    private final String token = "123456";
    private final String accessToken = "access token";
    private final String refreshToken = "refresh token";

    @Test
    void register_ShouldReturnCreatedUser() {
        UserCreationRequest request = new UserCreationRequest(email, password);
        UserCreationResult mockResult = new UserCreationResult(userId, email, expiresAt);
        when(createUserUseCase.execute(any(UserCreation.class))).thenReturn(mockResult);

        assertThat(
            mockMvcTester.post().uri("/api/v1/auth/register")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request))
        )
            .hasStatus(HttpStatus.CREATED)
            .bodyJson()
            .convertTo(UserCreationResponse.class)
            .satisfies(response -> {
                assertThat(response.userId()).isEqualTo(userId);
                assertThat(response.email()).isEqualTo(email);
                assertThat(response.expiresAt()).isEqualTo(expiresAt);
            });
    }

    @Test
    void register_WhenEmailAlreadyExists_ShouldReturnConflict() {
        String existingEmail = "existing@example.com";
        UserCreationRequest request = new UserCreationRequest(existingEmail, password);
        when(createUserUseCase.execute(any(UserCreation.class)))
            .thenThrow(UserExceptionFactory.emailAlreadyExists(existingEmail));

        assertThat(
            mockMvcTester.post().uri("/api/v1/auth/register")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request))
        )
            .hasStatus(HttpStatus.CONFLICT)
            .bodyJson()
            .extractingPath("$.detail").asString().contains(existingEmail);
    }

    @Test
    void confirmRegistration_ShouldReturnTokens_WhenRequestIsValid() {
        ConfirmRegistrationRequest request = new ConfirmRegistrationRequest(userId, token);
        TokenPair tokenPair = new TokenPair(accessToken, refreshToken);
        when(confirmRegistrationUseCase.execute(any(ConfirmRegistration.class))).thenReturn(tokenPair);

        assertThat(
            mockMvcTester.post().uri("/api/v1/auth/confirm-registration")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request))
        )
            .hasStatus(HttpStatus.OK)
            .bodyJson()
            .convertTo(AuthResponse.class)
            .satisfies(response -> {
                assertEquals(accessToken, response.accessToken());
                assertEquals(refreshToken, response.refreshToken());
            });
    }
}

File: src/test/kotlin/ru/jerael/booktracker/backend/web/controller/AuthControllerTest.kt
package ru.jerael.booktracker.backend.web.controller

import com.ninjasquad.springmockk.MockkBean
import io.mockk.Runs
import io.mockk.every
import io.mockk.just
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest
import org.springframework.context.annotation.Import
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.assertj.MockMvcTester
import ru.jerael.booktracker.backend.domain.exception.factory.UserExceptionFactory
import ru.jerael.booktracker.backend.domain.usecase.auth.*
import ru.jerael.booktracker.backend.domain.usecase.user.CreateUserUseCase
import ru.jerael.booktracker.backend.factory.auth.AuthDomainFactory
import ru.jerael.booktracker.backend.factory.auth.AuthWebFactory
import ru.jerael.booktracker.backend.factory.user.UserDomainFactory
import ru.jerael.booktracker.backend.factory.user.UserWebFactory
import ru.jerael.booktracker.backend.web.config.WebProperties
import ru.jerael.booktracker.backend.web.dto.auth.AuthResponse
import ru.jerael.booktracker.backend.web.dto.auth.ResendVerificationResponse
import ru.jerael.booktracker.backend.web.dto.user.UserCreationResponse
import ru.jerael.booktracker.backend.web.mapper.AuthWebMapper
import ru.jerael.booktracker.backend.web.mapper.UserWebMapper
import tools.jackson.databind.ObjectMapper

@WebMvcTest(AuthController::class)
@Import(UserWebMapper::class, AuthWebMapper::class)
class AuthControllerTest {
    
    @Autowired
    private lateinit var mockMvcTester: MockMvcTester
    
    @MockkBean
    private lateinit var webProperties: WebProperties
    
    @Autowired
    private lateinit var objectMapper: ObjectMapper
    
    @MockkBean
    private lateinit var createUserUseCase: CreateUserUseCase
    
    @MockkBean
    private lateinit var confirmRegistrationUseCase: ConfirmRegistrationUseCase
    
    @MockkBean
    private lateinit var loginUserUseCase: LoginUserUseCase
    
    @MockkBean
    private lateinit var refreshTokensUseCase: RefreshTokensUseCase
    
    @MockkBean
    private lateinit var logoutUserUseCase: LogoutUserUseCase
    
    @MockkBean
    private lateinit var resendVerificationUseCase: ResendVerificationUseCase
    
    @Test
    fun `register should return created User`() {
        val request = UserWebFactory.createUserCreationRequest()
        val result = UserDomainFactory.createUserCreationResult()
        every { createUserUseCase.execute(any()) } returns result
        
        assertThat(
            mockMvcTester.post().uri("/api/v1/auth/register")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request))
        )
            .hasStatus(HttpStatus.CREATED)
            .bodyJson()
            .convertTo(UserCreationResponse::class.java)
            .satisfies({ response ->
                assertEquals(result.userId, response.userId)
                assertEquals(result.email, response.email)
                assertEquals(result.expiresAt, response.expiresAt)
            })
    }
    
    @Test
    fun `when email already exists, register should return 409 CONFLICT`() {
        val existingEmail = "existing@example.com"
        val request = UserWebFactory.createUserCreationRequest(email = existingEmail)
        every { createUserUseCase.execute(any()) }
            .throws(UserExceptionFactory.emailAlreadyExists(existingEmail))
        
        assertThat(
            mockMvcTester.post().uri("/api/v1/auth/register")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request))
        )
            .hasStatus(HttpStatus.CONFLICT)
            .bodyJson()
            .extractingPath("$.detail").asString().contains(existingEmail)
    }
    
    @Test
    fun `when request is valid, confirmRegistration should return TokenPair`() {
        val request = AuthWebFactory.createConfirmRegistrationRequest()
        val tokenPair = AuthDomainFactory.createTokenPair()
        every { confirmRegistrationUseCase.execute(any()) } returns tokenPair
        
        assertThat(
            mockMvcTester.post().uri("/api/v1/auth/confirm-registration")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request))
        )
            .hasStatus(HttpStatus.OK)
            .bodyJson()
            .convertTo(AuthResponse::class.java)
            .satisfies({ response ->
                assertEquals(tokenPair.accessToken, response.accessToken)
                assertEquals(tokenPair.refreshToken, response.refreshToken)
            })
    }
    
    @Test
    fun `when request is valid, login should return TokenPair`() {
        val request = AuthWebFactory.createLoginRequest()
        val tokenPair = AuthDomainFactory.createTokenPair()
        every { loginUserUseCase.execute(any()) } returns tokenPair
        
        assertThat(
            mockMvcTester.post().uri("/api/v1/auth/login")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request))
        )
            .hasStatus(HttpStatus.OK)
            .bodyJson()
            .convertTo(AuthResponse::class.java)
            .satisfies({ response ->
                assertEquals(tokenPair.accessToken, response.accessToken)
                assertEquals(tokenPair.refreshToken, response.refreshToken)
            })
    }
    
    @Test
    fun `when request is valid, refresh should return TokenPair`() {
        val request = AuthWebFactory.createRefreshTokensRequest()
        val tokenPair = AuthDomainFactory.createTokenPair()
        every { refreshTokensUseCase.execute(any()) } returns tokenPair
        
        assertThat(
            mockMvcTester.post().uri("/api/v1/auth/refresh")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request))
        )
            .hasStatus(HttpStatus.OK)
            .bodyJson()
            .convertTo(AuthResponse::class.java)
            .satisfies({ response ->
                assertEquals(tokenPair.accessToken, response.accessToken)
                assertEquals(tokenPair.refreshToken, response.refreshToken)
            })
    }
    
    @Test
    fun `when request is valid, logout should return 200 OK`() {
        val request = AuthWebFactory.createLogoutRequest()
        every { logoutUserUseCase.execute(any()) } just Runs
        
        assertThat(
            mockMvcTester.post().uri("/api/v1/auth/logout")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request))
        )
            .hasStatus(HttpStatus.OK)
    }
    
    @Test
    fun `when request is valid, resendCode should return ResendVerificationResponse`() {
        val request = AuthWebFactory.createResendVerificationRequest()
        val result = AuthDomainFactory.createResendVerificationResult()
        every { resendVerificationUseCase.execute(any()) } returns result
        
        assertThat(
            mockMvcTester.post().uri("/api/v1/auth/resend-code")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request))
        )
            .hasStatus(HttpStatus.OK)
            .bodyJson()
            .convertTo(ResendVerificationResponse::class.java)
            .satisfies({ response ->
                assertEquals(result.userId, response.userId)
                assertEquals(result.email, response.email)
                assertEquals(result.expiresAt, response.expiresAt)
            })
    }
}